crypto/threads_common.c

```diff
@@ -0,0 +1,451 @@
+ * be created. This function allows each thread to maintain separate data
+ * for different library contexts under a shared key identifier.
+ *
+ * @param id
+ *        The thread-local key ID to associate the data with.
+ *
+ * @param ctx
+ *        Pointer to the `OSSL_LIB_CTX` used as a secondary key for storing
+ *        the data.
+ *
+ * @param data
+ *        Pointer to the user-defined context-specific data to store.
+ *
+ * @return 1 on success, or 0 if allocation or initialization fails.
+ */
+int CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_KEY_ID id,
+                               OSSL_LIB_CTX *ctx, void *data)
+{
+    SPARSE_ARRAY_OF(MASTER_KEY_ENTRY) *mkey;
+    MASTER_KEY_ENTRY *entry;
+    CTX_TABLE_ENTRY *ctxd;
+
+    /*
+     * Make sure our master key is initalized
+     * See notes above on the use of CRYPTO_THREAD_run_once here
+     */
+    if (!CRYPTO_THREAD_run_once(&master_once, init_master_key))
+        return 0;
+
+    /*
+     * Get our local master key data, which will be
+     * a sparse array indexed by the id parameter
+     */
+    mkey = CRYPTO_THREAD_get_local(&master_key);
+    if (mkey == NULL) {
+        /*
+         * we didn't find one, but thats ok, just initalize it now
+         */
+        mkey = ossl_sa_MASTER_KEY_ENTRY_new();
+        if (mkey == NULL)
+            return 0;
+        /*
+         * make sure to assign it to our master key thread-local storage
+         */
+        CRYPTO_THREAD_set_local(&master_key, mkey);
+    }
+
+    /*
+     * Find the entry that we are looking for using our id index
+     */
+    entry = ossl_sa_MASTER_KEY_ENTRY_get(mkey, id);
+    if (entry == NULL) {
+
+        /*
+         * Didn't find it, thats ok, just add it now
+         */
+        entry = OPENSSL_zalloc(sizeof(MASTER_KEY_ENTRY));
+        if (entry == NULL)
+            return 0;
+        entry->ctx_table = ossl_sa_CTX_TABLE_ENTRY_new();
+        if (entry->ctx_table == NULL) {
+            OPENSSL_free(entry);
+            return 0;
+        }
+        /*
+         * Assign it to the appropriate entry in the master key sparse array
+         */
+        ossl_sa_MASTER_KEY_ENTRY_set(mkey, id, entry);
+    }
+
+    /*
+     * Now go look up our per context entry, using the OSSL_LIB_CTX pointer
+     * that we've been provided.  Note we cast the pointer to a uintptr_t so
+     * as to use it as an index in the sparse array
+     */
+    ctxd = ossl_sa_CTX_TABLE_ENTRY_get(entry->ctx_table, (uintptr_t)ctx);
+    if (ctxd == NULL) {
+        /*
+         * No entry for this context, build one
+         */
+        ctxd = OPENSSL_zalloc(sizeof(CTX_TABLE_ENTRY));
+        if (ctxd == NULL)
+            return 0;
+        /*
+         * Assign to the entry in the table so that we can find it later
+         */
+        ossl_sa_CTX_TABLE_ENTRY_set(entry->ctx_table, (uintptr_t)ctx, ctxd);
+    }
+    /*
+     * Lastly assign our data pointer
+     */
+    ctxd->ctx_data = data;
```

Comment:
Shall we fail if `ctxd->ctx_data` is not `NULL` so we can give a caller chance to avoid a memory leak (accident overwrite of existing key)? Perhaps more changes would be required if we will go this way.

Reply:
Normally the call pattern to update a thread-local pointer looks like this:
```
tmp = CRYPTO_THREAD_get_local[_ex](key, [ctx]);
CRYPTO_THREAD_set_local[_ex](key, [ctx], newptr);
OPENSSL_free(tmp);
```

If we were to introduce a failure on non-null then the calling pattern would have to be something like:
```
tmp = CRYPTO_THREAD_get_local[_ex](key, [ctx]);
CRYPTO_THREAD_set_local[_ex](key, [ctx], NULL);
CRYPTO_THREAD_set_local[_ex](key, [ctx], newptr);
OPENSSL_free(tmp);
```
which seems less efficient.

Given that asan will catch any leaks, I think we're ok as we are on this account.